src/cli-sdk/src/pack-tarball.ts

```diff
@@ -16,181 +18,316 @@ export type PackTarballResult = {
   shasum?: string
 }
 
+/**
+ * Replace workspace: and catalog: specs with actual versions
+ * @param {Manifest} manifest_ - The manifest to process
+ * @param {LoadedConfig} config - The loaded configuration containing project root, monorepo, and catalog data
+ * @returns {Manifest} The manifest with replaced specs
+ */
+const replaceWorkspaceAndCatalogSpecs = (
+  manifest_: Manifest,
+  config: LoadedConfig,
+): Manifest => {
+  // Create a deep copy of the manifest to avoid modifying the original
+  const manifest = JSON.parse(JSON.stringify(manifest_)) as Manifest
+
+  // Get workspace and catalog configuration from config
+  const { monorepo, catalog = {}, catalogs = {} } = config.options
+
+  // Process dependency types
+  const depTypes = [
+    'dependencies',
+    'devDependencies',
+    'optionalDependencies',
+    'peerDependencies',
+  ] as const
+
+  for (const depType of depTypes) {
+    const deps = manifest[depType]
+    /* c8 ignore next */
+    if (!deps || typeof deps !== 'object') continue
+
+    const depsObj = deps as Record<string, unknown>
+    for (const [depName, depSpec] of Object.entries(depsObj)) {
```

Comment:
future refactor idea here, let's get the `./dependencies.ts` module out of `@vltpkg/graph` and into its own workspace (so that we don't need to have a dependency on the entire graph library just for reusing this kind of functionality) and reuse the `getDependencies` method https://github.com/vltpkg/vltpkg/blob/5a685ed51dd9d3986104440320ab3625bd1c1c5c/src/graph/src/dependencies.ts#L188-L210

it's already reused in a bunch of places and I think it makes sense to have a standard way of performing this dance

Reply:
it needs a little bit more abstraction though as it should take only the manifest instead of relying on the node but anyways, just wanted to flag as a possible directoin in the future, I don't think it's worth investing time on it rn

Reply:
I was also thinking it would be nice to have a method that read a manifest and would optionally do all the replacements for you. That could live in the same workspace maybe?

Reply:
ooohhh maybe 🤔 one thing to also keep in mind is the `normalizeManifest` stuff that recently got beffier with @darcyclarke landing more functions to normalize `contributors` and `maintainers`